mempool/src/pool/tests.rs

```diff
@@ -0,0 +1,1787 @@
+    let locktime = 0;
+    let outpoint_source_id = OutPointSourceId::Transaction(parent_id);
+    let child = tx_spend_input(
+        &mempool,
+        TxInput::new(
+            outpoint_source_id,
+            0,
+            InputWitness::NoSignature(Some(DUMMY_WITNESS_MSG.to_vec())),
+        ),
+        None,
+        flags,
+        locktime,
+    )?;
+    let child_id = child.get_id();
+    mock_clock.set(DEFAULT_MEMPOOL_EXPIRY + Duration::new(1, 0));
+
+    assert!(matches!(
+        mempool.add_transaction(child),
+        Err(Error::TxValidationError(
+            TxValidationError::DescendantOfExpiredTransaction
+        ))
+    ));
+
+    assert!(!mempool.contains_transaction(&parent_id));
+    assert!(!mempool.contains_transaction(&child_id));
+    Ok(())
+}
+
+#[test]
+fn mempool_full() -> anyhow::Result<()> {
+    logging::init_logging::<&str>(None);
+    let mut mock_usage = MockGetMemoryUsage::new();
+    mock_usage
+        .expect_get_memory_usage()
+        .times(1)
+        .return_const(MAX_MEMPOOL_SIZE_BYTES + 1);
+
+    let chain_state = ChainStateMock::new();
+    let mut mempool = MempoolImpl::create(chain_state, SystemClock, mock_usage);
+
+    let tx = TxGenerator::new().generate_tx(&mempool)?;
+    log::debug!("mempool_full: tx has is {}", tx.get_id().get());
+    assert!(matches!(
+        mempool.add_transaction(tx),
+        Err(Error::MempoolFull)
+    ));
+    Ok(())
+}
+
+#[test]
+fn no_empty_bags_in_descendant_score_index() -> anyhow::Result<()> {
+    let mut mempool = setup();
+
+    let num_inputs = 1;
+    let num_outputs = 100;
+    let fee = get_relay_fee_from_tx_size(estimate_tx_size(num_inputs, num_outputs));
+    let parent = TxGenerator::new()
+        .with_num_inputs(num_inputs)
+        .with_num_outputs(num_outputs)
+        .with_fee(Amount::from_atoms(fee))
+        .generate_tx(&mempool)?;
+    let parent_id = parent.get_id();
+
+    let outpoint_source_id = OutPointSourceId::Transaction(parent.get_id());
+    mempool.add_transaction(parent)?;
+    let num_child_txs = num_outputs;
+    let flags = 0;
+    let locktime = 0;
+    let txs = (0..num_child_txs)
+        .into_iter()
+        .map(|i| {
+            tx_spend_input(
+                &mempool,
+                TxInput::new(
+                    outpoint_source_id.clone(),
+                    u32::try_from(i).unwrap(),
+                    InputWitness::NoSignature(Some(DUMMY_WITNESS_MSG.to_vec())),
+                ),
+                Amount::from_atoms(fee + u128::try_from(i).unwrap()),
+                flags,
+                locktime,
+            )
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+    let ids = txs.iter().map(|tx| tx.get_id()).collect::<Vec<_>>();
+
+    for tx in txs {
+        mempool.add_transaction(tx)?;
+    }
+
+    mempool.drop_transaction(&parent_id);
+    for id in ids {
+        mempool.drop_transaction(&id)
+    }
+    assert!(mempool.store.txs_by_descendant_score.is_empty());
+    Ok(())
+}
```

Comment:
Here's a testing idea. Introduce a new method to `MempoolStore`, called say `assert_valid`, that is enabled for testing (does nothing otherwise) and contains a number of potentially expensive integrity checks:

* Every transaction ID form `txs_by_id` occurs in `txs_by_descendant_score` exactly once
* For each entry, each child has the entry as a parent
* And so on, there are many more of these I can imagine

Then sprinkle calls to it throughout the tests or into implementation of methods that are expected to preserve these invariants.

Reply:
Opened an issue for this :-D.